File: phylo2vec/base/to_newick.py
"""
Methods to convert a Phylo2Vec vector to a Newick-format string.
"""

import numba as nb
import numpy as np

# Numba tuple type
value_type = nb.types.UniTuple(nb.types.int64, 2)


@nb.njit(cache=True)
def _get_pairs(v):
    pairs = []

    for i in range(len(v) - 1, -1, -1):
        next_leaf = i + 1
        if v[i] <= i:
            # If v[i] <= i, it's an easy BD
            # We now that the next pair to add now is (v[i], next_leaf)
            # (as the branch leading to v[i] gives birth to the next_leaf)
            # Why pairs.insert(0)? Let's take an example with [0, 0]
            # We initially have (0, 1), but 0 gives birth to 2 afterwards
            # So the "shallowest" pair is (0, 2)
            pairs.append((v[i], next_leaf))

    for j in range(1, len(v)):
        next_leaf = j + 1
        if v[j] == 2 * j:
            # 2*j = extra root ==> pairing = (0, next_leaf)
            pairs.append((0, next_leaf))
        elif v[j] > j:
            # If v[i] > i, it's not the branch leading v[i] that gives birth but an internal branch
            # Remark 1: it will not be the "shallowest" pair, so we do not insert it at position 0
            # len(pairs) = number of pairings we did so far
            # So what v[i] - len(pairs) gives us is the depth of the next pairing
            # And pairs[v[i] - len(pairs) - 1][0] is a node that we processed beforehand
            # which is deeper than the branch v[i]
            index = v[j] - 2 * j
            pairs.insert(index, (pairs[index - 1][0], next_leaf))

    return pairs


@nb.njit(cache=True)
def _get_ancestry(v):
    """
    Get the "ancestry" of v (see "Returns" paragraph)

    v[i] = which BRANCH we do the pairing from

    The initial situation looks like this:
                      R
                      |
                      | --> branch 2
                    // \\
      branch 0 <-- //   \\  --> branch 1
                   0     1

    For v[1], we have 3 possible branches too choose from.
    v[1] = 0 or 1 indicates that we branch out from branch 0 or 1, respectively.
    The new branch yields leaf 2 (like in ordered trees)

    v[1] = 2 is somewhat similar: we create a new branch from R that yields leaf 2

    Parameters
    ----------
    v : numpy.array
        Phylo2Vec vector

    Returns
    -------
    ancestry : numpy.array
        Ancestry matrix
        1st column: child 1
        2nd column: child 2
        3rd column: parent node
    """
    pairs = _get_pairs(v)

    # We have our pairs, we can now build our ancestry
    # Matrix with 3 columns: child1, child2, parent
    ancestry = np.zeros((len(pairs), 3), dtype=np.int16)

    # Dictionary to keep track of the following relationship: child->highest parent
    parents = nb.typed.Dict.empty(key_type=nb.types.int64, value_type=nb.types.int64)

    # Dictionary to keep track of siblings (i.e., sister nodes)
    # siblings = nb.typed.Dict.empty(key_type=nb.types.int64, value_type=nb.types.int64)

    # Leaves are number 0, 1, ..., n_leaves - 1, so the next parent is n_leaves
    next_parent = len(v) + 1

    for i, pair in enumerate(pairs):
        child1, child2 = pair

        parent_child1 = parents.get(child1, child1)
        parent_child2 = parents.get(child2, child2)

        ancestry[i, :] = [parent_child1, parent_child2, next_parent]

        # Change the parents of the current children
        parents[child1] = next_parent
        parents[child2] = next_parent

        next_parent += 1

    return ancestry


@nb.njit
def _build_newick(ancestry):
    """Build a Newick string from an "ancestry" array

    The input should always be 3-dimensional with the following format:
    1st column: child 1
    2nd column: child 2
    3rd column: parent node

    The matrix is processed such that we iteratively write a Newick string
    to describe the tree.

    Parameters
    ----------
    ancestry : numpy.ndarray
        "Ancestry" array of size (n_leaves - 1, 3)

    Returns
    -------
    newick : str
        Newick string
    """

    # TODO: drop the ancestry matrix form?
    ancestry_dict = nb.typed.Dict.empty(key_type=nb.types.int64, value_type=value_type)

    for c1, c2, p in ancestry:
        ancestry_dict[p] = (c1, c2)

    newick = f"{_build_newick_inner(ancestry[-1][-1], ancestry_dict)};"

    return newick


@nb.njit
def _build_newick_inner(node, ancestry_dict):
    if node in ancestry_dict:
        c1, c2 = ancestry_dict.pop(node)
        return "".join(
            (
                "(",
                _build_newick_inner(c1, ancestry_dict),
                _build_newick_inner(c2, ancestry_dict),
                ")",
                f"{node}",
            )
        )

        #    ( "("
        #     + _build_newick_inner(c1, ancestry_dict)
        #     + ","
        #     + _build_newick_inner(c2, ancestry_dict)
        #     + ")"
        #     + f"{node}"
        # )
    else:
        return f"{node}"


@nb.njit
def to_newick(v):
    """Recover a rooted tree (in Newick format) from a Phylo2Vec v

    Parameters
    ----------
    v : numpy.array
        Phylo2Vec vector

    Returns
    -------
    newick : str
        Newick tree
    """
    ancestry = _get_ancestry(v)

    newick = _build_newick(ancestry)

    return newick

File: phylo2vec/src/to_newick.rs
use crate::avl::{AVLTree, Pair};

// A type alias for the Ancestry type, which is a vector of tuples representing (child1, child2, parent)
type Ancestry = Vec<(usize, usize, usize)>;

pub fn _get_pairs(v: &Vec<usize>) -> Vec<(usize, usize)> {

    let num_of_leaves: usize = v.len();
    let mut pairs: Vec<(usize, usize)> = Vec::with_capacity(num_of_leaves);

    // First loop (reverse iteration)
    for i in (0..num_of_leaves).rev() {
        /*
        If v[i] <= i, it's like a birth-death process.
        The next pair to add is (v[i], next_leaf) as the branch leading to v[i]
        gives birth to next_leaf.
        */
        let next_leaf: usize = i + 1;
        let pair = (v[i], next_leaf);
        if v[i] <= i {
            pairs.push(pair);
        }
    }

    // Second loop
    for j in 1..num_of_leaves {
        let next_leaf = j + 1;
        if v[j] == 2 * j {
            // 2 * j = extra root ==> pairing = (0, next_leaf)
            let pair = (0, next_leaf);
            pairs.push(pair);
        } else if v[j] > j {
            /*
            If v[j] > j, it's not the branch leading to v[j] that gives birth,
            but an internal branch. Insert at the calculated index.
            */
            let index = pairs.len() + v[j] - 2 * j;
            let new_pair = (pairs[index - 1].0, next_leaf);
            pairs.insert(index, new_pair);
        }
    }

    pairs
}


pub fn _get_pairs_avl(v: &Vec<usize>) -> Vec<Pair> {
    // AVL tree implementation of get_pairs
    let k = v.len();
    let mut avl_tree = AVLTree::new();
    avl_tree.insert(0, (0, 1));

    for i in 1..k {
        let next_leaf = i + 1;
        if v[i] <= i {
            avl_tree.insert(0, (v[i], next_leaf));
        } else {
            let index = v[i] - next_leaf;
            let pair = AVLTree::lookup(&avl_tree, index);
            avl_tree.insert(index + 1, (pair.0, next_leaf));
        }
    }

    avl_tree.get_pairs()
}



/// Get the ancestry of the Phylo2Vec vector
pub fn _get_ancestry(v: &Vec<usize>) -> Ancestry {
    let pairs = _get_pairs(&v);
    let num_of_leaves = v.len();
    // Initialize Ancestry with capacity `k`
    let mut ancestry: Ancestry = Vec::with_capacity(num_of_leaves);
    // Keep track of child->highest parent relationship
    let mut parents: Vec<isize> = vec![-1; 2 * num_of_leaves + 1];

    for i in 0..num_of_leaves {
        let (c1, c2) = pairs[i];

        let parent_of_child1 = if parents[c1] != -1 {
            parents[c1] as usize
        } else {
            c1
        };
        let parent_of_child2 = if parents[c2] != -1 {
            parents[c2] as usize
        } else {
            c2
        };

        // Next parent
        let next_parent = (num_of_leaves + i + 1) as isize;
        ancestry.push((parent_of_child1, parent_of_child2, next_parent as usize));

        // Update the parents of current children
        parents[c1] = next_parent;
        parents[c2] = next_parent;
    }

    ancestry
}

// The recursive function that builds the Newick string
fn _build_newick_recursive_inner(p: usize, ancestry: &Ancestry) -> String {
    let leaf_max = ancestry.len();

    // Extract the children (c1, c2) and ignore the parent from the ancestry tuple
    let (c1, c2, _) = ancestry[p - leaf_max - 1];

    // Recursive calls for left and right children, checking if they are leaves or internal nodes
    let left = if c1 > leaf_max {
        _build_newick_recursive_inner(c1, ancestry)
    } else {
        c1.to_string() // It's a leaf node, just convert to string
    };

    let right = if c2 > leaf_max {
        _build_newick_recursive_inner(c2, ancestry)
    } else {
        c2.to_string() // It's a leaf node, just convert to string
    };

    // Create the Newick string in the form (left, right)p
    format!("({},{}){}", left, right, p)
}

// The main function to build the Newick string from the ancestry
pub fn _build_newick(ancestry: &Ancestry) -> String {
    // Get the root node, which is the parent value of the last ancestry element
    let root = ancestry.last().unwrap().2;

    // Build the Newick string starting from the root, and append a semicolon
    format!("{};", _build_newick_recursive_inner(root, ancestry))
}

/// Recover a rooted tree (in Newick format) from a Phylo2Vec vector
pub fn to_newick(v: Vec<usize>) -> String {
    let ancestry = _get_ancestry(&v);
    _build_newick(&ancestry)
}
